homeassistant/components/energyzero/services.py

```diff
@@ -0,0 +1,126 @@
+"""The EnergyZero services."""
+from __future__ import annotations
+
+from datetime import date, datetime
+from enum import Enum
+from functools import partial
+from typing import Final
+
+from energyzero import Electricity, Gas, VatOption
+import voluptuous as vol
+
+from homeassistant.core import (
+    HomeAssistant,
+    ServiceCall,
+    ServiceResponse,
+    SupportsResponse,
+)
+from homeassistant.exceptions import ServiceValidationError
+from homeassistant.util import dt as dt_util
+
+from .const import DOMAIN
+from .coordinator import EnergyZeroDataUpdateCoordinator
+
+ATTR_START: Final = "start"
+ATTR_END: Final = "end"
+ATTR_INCL_VAT: Final = "incl_vat"
+
+GAS_SERVICE_NAME: Final = "get_gas_prices"
+ENERGY_SERVICE_NAME: Final = "get_energy_prices"
+SERVICE_SCHEMA: Final = vol.Schema(
+    {
+        vol.Required(ATTR_INCL_VAT): bool,
+        vol.Optional(ATTR_START): str,
+        vol.Optional(ATTR_END): str,
+    }
+)
+
+
+class PriceType(Enum):
+    """Type of price."""
+
+    ENERGY = "energy"
+    GAS = "gas"
+
+
+def __get_date(date_input: str | None) -> date | datetime:
+    """Get date."""
+    if not date_input:
+        return dt_util.now().date()
+
+    if value := dt_util.parse_datetime(date_input):
+        return value
+
+    raise ServiceValidationError(
+        translation_domain=DOMAIN,
+        translation_key="invalid_date",
+        translation_placeholders={
+            "date": date_input,
+        },
+    )
+
+
+def __serialize_prices(prices: Electricity | Gas) -> ServiceResponse:
+    """Serialize prices."""
+    return {
+        "prices": [
+            {
+                key: str(value) if isinstance(value, datetime) else value
+                for key, value in timestamp_price.items()
+            }
+            for timestamp_price in prices.timestamp_prices
+        ]
+    }
+
+
+async def __get_prices(
+    call: ServiceCall,
+    *,
+    coordinator: EnergyZeroDataUpdateCoordinator,
+    price_type: PriceType,
+) -> ServiceResponse:
+    start = __get_date(call.data.get(ATTR_START))
+    end = __get_date(call.data.get(ATTR_END))
+
+    vat = VatOption.INCLUDE
+
+    if call.data.get(ATTR_INCL_VAT) is False:
+        vat = VatOption.EXCLUDE
+
+    data: Electricity | Gas
+
+    if price_type == PriceType.GAS:
+        data = await coordinator.energyzero.gas_prices(
+            start_date=start,
+            end_date=end,
+            vat=vat,
+        )
+    else:
+        data = await coordinator.energyzero.energy_prices(
+            start_date=start,
+            end_date=end,
+            vat=vat,
+        )
+
+    return __serialize_prices(data)
+
+
+async def async_register_services(
```

Comment:
```suggestion
@callback
def async_register_services(
```

See https://developers.home-assistant.io/docs/asyncio_categorizing_functions/ for explanations

Reply:
How important is this? The EasyEnergy PR that is based on this PR does not have the callback and is merged: https://github.com/home-assistant/core/pull/105175/files#diff-8162f13dcf871fea4e83ae6cb8fadf9adfe6ddcaa58e6da14edb5478041106a0R112

Reply:
https://github.com/home-assistant/core/pull/105947 will fix it in the above PR